app/src/main/java/org/example/BoardPosition.java

```diff
@@ -0,0 +1,119 @@
+package org.example;
+
+public class BoardPosition {
+    public final int i;
+    public final int j;
+
+    private BoardPosition(int i, int j) {
+        this.i = i;
+        this.j = j;
+    }
+
+    // Singleton class to represent an invalid position. Useful for chaining moves:
+    public static class InvalidBoardPosition extends BoardPosition {
+        private static final InvalidBoardPosition mInstance = new InvalidBoardPosition();
+
+        private InvalidBoardPosition() {
+            super(-1, -1);
+        }
+
+        public static InvalidBoardPosition getInstance() {
+            return mInstance;
+        }
+
+        @Override
+        public boolean isValid() {
+            return false;
+        }
+
+        @Override
+        public BoardPosition right() {
+            return this;
+        }
+
+        @Override
+        public BoardPosition left() {
+            return this;
+        }
+
+        @Override
+        public BoardPosition ahead() {
+            return this;
+        }
+
+        @Override
+        public BoardPosition behind() {
+            return this;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return this == obj;
+        }
+    }
+
+    public static BoardPosition fromCoords(int i, int j) {
+        if (isValid(i, j)) {
+            return new BoardPosition(i, j);
+        } else {
+            return InvalidBoardPosition.getInstance();
+        }
+
+    }
+
+    public boolean isValid() {
+        return i >= 0 && i < 8 && j >= 0 && j < 8;
+    }
+
+    private static boolean isValid(int i, int j) {
+        return i >= 0 && i < 8 && j >= 0 && j < 8;
+    }
```

Comment:
This defeats the purpose of having BoardPosition. It should always be valid if it exists.